### PR TITLE
Initial open-source implementation

### PR DESCRIPTION
**Pull request by simu**
## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Update the documentation.
- [x] Update the ./CHANGELOG.md.

### CODE REVIEW

File: component/prometheus.jsonnet [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        memory: '2Gi',
+        cpu: '1000m',
+      },
+      limits+: {
+        memory: '6Gi',
+        cpu: '2000m',
+      },
+    },
+    storage+: {
+      volumeClaimTemplate+: {
```
Comment (srueg) [inline component/prometheus.jsonnet:74]: Looks like this is already defined in `defaults.yml`?

File: class/defaults.yml [hunk 2]
```diff
@@ -1,3 +1,54 @@
...
+    # alertmanagerConfig is rendered as is into the alertmanager config secret
+    alertmanagerConfig:
+      receivers:
+        # blackhole receiver for alerts we don't care about
+        - name: devnull
+      # If the user doesn't configure a receiver, send everything to
+      # devnull. This allows Alertmanager to start even if the user doesn't
+      # configure any alert receivers.
+      route:
+        receiver: devnull
```
Comment (srueg) [inline class/defaults.yml:38]: I'd propose to set the following additional defaults:
```suggestion
        receiver: devnull
        group_wait: 0s
        group_interval: 5s
        repeat_interval: 10m
```

File: component/prometheus.jsonnet [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    nodeSelector+: {
+      'kubernetes.io/os': 'linux',
+    },
+    podMonitorNamespaceSelector+: nsSelector,
+    ruleNamespaceSelector+: nsSelector,
+    serviceMonitorNamespaceSelector+: nsSelector,
+    ruleSelector+: {
+      matchLabels: matchLabels,
+    },
+    replicas: 1,
```
Comment (srueg) [inline component/prometheus.jsonnet:62]: It might make more sense to have most of these default values in `default.yaml` to have them in one place, wdyt?

Additionally I'd suggest to set the following additional defaults:
```
evaluationInterval: 5s
scrapeInterval: 10s
```

File: component/prometheus.jsonnet [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+  apiVersion: 'monitoring.coreos.com/v1',
+  kind: 'Prometheus',
+  metadata: {
+    name: name,
+    namespace: namespace,
+    labels: labels,
+  },
+  spec: {
+    externalLabels: {
+      customer: inv.parameters.customer.name,
```
Comment (srueg) [inline component/prometheus.jsonnet:39]:
```suggestion
      tenant: inv.parameters.cluster.tenant,
```

**Review by srueg — CHANGES_REQUESTED** [review]

File: component/prometheus.jsonnet (hunk 3 above)
Comment (simu) [reply component/prometheus.jsonnet:62]: I was thinking about having most of the default values in `default.yml`. The only downside I can see is that somebody could do

```
parameters:
  rancher_monitoring:
    prometheus: ~
```

followed by

```
parameters:
  rancher_monitoring:
    prometheus:
      [their customization]
```

which could possibly lead to non-functional configurations. However, I haven't checked whether the fields we're setting currently are actually required.

tl;dr: I'm fine with moving the defaults to `class/defaults.yml`, if we accept that users may be able to create non-functional Prometheus/Alertmanager objects

**Review by srueg — APPROVED** [review]

File: component/alertmanager.jsonnet [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+local kap = import 'lib/kapitan.libjsonnet';
+local kube = import 'lib/kube.libjsonnet';
+local inv = kap.inventory();
+local params = inv.parameters.rancher_monitoring;
+local namespace = params.namespace;
+local instance = params.alertmanagerInstance;
+local name = 'alertmanager-' + instance;
+
+local labels = {
+  'app.kubernetes.io/name': 'alertmanager',
```
Comment (corvus-ch) [inline component/alertmanager.jsonnet:11]:
```suggestion
  'app.kubernetes.io/name': name,
```

Not sure about this one.

File: component/alertmanager.jsonnet [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+};
+
+local sa = kube.ServiceAccount(name) {
+  metadata+: {
+    namespace: namespace,
+    labels+: labels,
+  },
+};
+
+local alertmanager = {
```
Comment (corvus-ch) [inline component/alertmanager.jsonnet:28]: Is there a specific reason to no use the alertmanager from `kube-prometheus`?

File: component/prometheus.jsonnet [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+};
+
+local nsSelector = {
+  matchLabels: {
+    SYNMonitoring: 'main',
+  },
+};
+
+// The actual prometheus resource for the prometheus operator
+local prometheus = {
```
Comment (corvus-ch) [inline component/prometheus.jsonnet:29]: Again, we might be able to profit from `kube-prometheus`.

File: component/rules.jsonnet [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    ],
+  },
+};
+
+local kp =
+  (import 'kube-prometheus/kube-prometheus.libsonnet') +
+  (import 'kube-prometheus/kube-prometheus-managed-cluster.libsonnet') +
+  filterRules +
+  alterRules +
+  additionalRules +
```
Comment (corvus-ch) [inline component/rules.jsonnet:110]: Adding `additionalRules` last, makes this hard coded. `filterRules` and `alterRules` can not influence it. Is this intentional?

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 9]
```diff
@@ -8,8 +8,173 @@ The parent key for all of the following parameters is `rancher_monitoring`.
...
+----
+interval: 10s
+scrape_timeout: 10s
+----
+
+Configure the scrape interval and timeout for the Prometheus job which federates metrics from the Rancher Prometheus instance in `cattle-prometheus`.
+
+Users should ensure that the `scrape_timeout` is lower than the `interval`, as there's no validation logic in the component.
+
+== `exclude_rules`
```
Comment (corvus-ch) [inline docs/modules/ROOT/pages/references/parameters.adoc:168]: Consider to alinge this with https://github.com/appuio/component-openshift4-monitoring/blob/master/class/defaults.yml#L56.

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 10]
```diff
@@ -8,8 +8,173 @@ The parent key for all of the following parameters is `rancher_monitoring`.
...
+[horizontal]
+type:: list
+default:: `[]`
+
+A list of alert rules which should be completely disabled on the cluster.
+Rules are listed as dicts with keys `group` and `name`.
+The `group` key is matched against the `kube-prometheus` alert rule group name.
+The `name` key is used to filter out alerts with matching name from the given group.
 
 == Example
```
Comment (corvus-ch) [inline docs/modules/ROOT/pages/references/parameters.adoc:179]: Please come up with a more elaborate example.

**Review by corvus-ch — CHANGES_REQUESTED** [review]
We could make use of the Aletermanager and Prometheus setup from `kube-prometheus` (excluding the operator itself). This woul probably reduce a lot of lines within this component. Please consider to do this.

As for the requested change: the example in the documentation can not be left that barebone.

**Comment by simu** [comment]
> We could make use of the Aletermanager and Prometheus setup from `kube-prometheus` (excluding the operator itself). This woul probably reduce a lot of lines within this component. Please consider to do this.

This is out-of-scope for the initial open-source version of the component, in the interest of keeping the manifest diffs minimal when switching to the open-source component.

File: component/alertmanager.jsonnet (hunk 6 above)
Comment (simu) [reply component/alertmanager.jsonnet:28]: For the initial open-source version I want to keep the resulting manifest diffs as small as possible to allow migrating to the open-source component with minimal effort.

We can switch to alertmanager from `kube-prometheus` in a follow-up issue/PR at any time if we think it's beneficial.

File: component/prometheus.jsonnet (hunk 7 above)
Comment (simu) [reply component/prometheus.jsonnet:29]: see comment above

File: component/alertmanager.jsonnet (hunk 5 above)
Comment (simu) [reply component/alertmanager.jsonnet:11]: According to https://syn.tools/syn/references/labels-guide.html which has the information from the Kubernetes docs:

> `app.kubernetes.io/name` -- The name of the application -- `steward`

Given that information, `alertmanager` should be the correct value for the `app.kubernetes.io/name` label

File: component/rules.jsonnet (hunk 8 above)
Comment (simu) [reply component/rules.jsonnet:108]: Adjusted, thanks for pointing that out.

(side note: alterRules probably shouldn't be used to modify rules which are defined in the same jsonnet file, but that's another discussion)

**Review by corvus-ch — APPROVED** [review]
